### PR TITLE
[core] Fix HTTP redirect handling

### PR DESCRIPTION
**Pull request by coletdjnz**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

ADD DESCRIPTION HERE

Backport from #2861 

https://github.com/yt-dlp/yt-dlp/commit/afac4caa7db30804bebac33e53c3cb0237958224 never properly fixed the issue (and was not tested, which was part of the problem):
- Did not handle 307/308 for POST (https://github.com/python/cpython/issues/91306)
- Removed payload regardless if method changed or not (so POST to POST would strip payload. It should only remove if method changed i.e. POST to GET)
- Did not support PUT and other methods for redirect.

This PR fixes the above issues. It aligns the redirect handling with the latest recommended RFC standard/what common browsers tend to do (see #2861 for more full description of this).

I have also back-ported some of the networking tests, including the HTTP redirect test.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [X] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🧹🐛🎨

Improve redirection handling and code style in `yt_dlp/utils/_utils.py` and remove obsolete code and tests. This pull request enhances the functionality, readability, and maintainability of the codebase.

> _`yt_dlp` cleans up_
> _Redirection bug is fixed_
> _Autumn leaves unused code_

### Walkthrough
*  Simplify and fix redirection logic in `YoutubeDLRedirectHandler` class (`[link](https://github.com/yt-dlp/yt-dlp/pull/7094/files?diff=unified&w=0#diff-feda8f56946dc048e754111850baaef0eec4b9f8bbc2d3f04b1a785626ea5c0eL1682-R1683)`, `[link](https://github.com/yt-dlp/yt-dlp/pull/7094/files?diff=unified&w=0#diff-feda8f56946dc048e754111850baaef0eec4b9f8bbc2d3f04b1a785626ea5c0eL1696-R1717)`)

</details>

### CODE REVIEW

File: test/test_networking.py
Comment (coletdjnz) [inline test/test_networking.py:291]: test that the no compression header is not present

File: test/test_networking.py
Comment (coletdjnz) [inline test/test_networking.py:8]: move imports below

File: test/test_networking.py
Comment (Grub4K) [inline test/test_networking.py:272]:
```suggestion
            r = sanitized_Request(f'http://localhost:{self.http_port}/headers', data=urlencode_postdata({'test': 'test'}))
```

File: test/test_networking.py
Comment (Grub4K) [inline test/test_networking.py:266]:
```suggestion
            r = sanitized_Request(f'https://localhost:{self.https_port}/headers', data=urlencode_postdata({'test': 'test'}))
```

File: test/test_networking.py
Comment (Grub4K) [inline test/test_networking.py:255]: Maybe use a loop going over all methods to test them not getting modified?

File: test/test_networking.py
Comment (Grub4K) [inline test/test_networking.py:235]:
```suggestion
                res = ydl.urlopen(sanitized_Request(
                    f'http://127.0.0.1:{self.http_port}/redirect_{redirect_status}', method=method, data=data))
```

File: test/test_networking.py
Comment (Grub4K) [inline test/test_networking.py:218]: If that is the tests purpose, shouldn't `req.url` be checked to be the same?

File: test/test_networking.py
Comment (Grub4K) [inline test/test_networking.py:144]:
```suggestion
            compressed = buf.getvalue() + b'trailing garbage'
            self.send_header('Content-Length', str(len(compressed)))
            self.end_headers()
            self.wfile.write(compressed)
```

File: test/test_networking.py
Comment (Grub4K) [inline test/test_networking.py:146]:
```suggestion
            new_url = f'http://127.0.0.1:{http_server_port(self.server)}/中文.html'
```

**Review by Grub4K — COMMENTED** [review]
```py
with FakeYDL(...) as ydl:
    ...
```
Can be rewritten as `ydl = FakeYDL(...)`, saving one indentation across the file, since we are currently only using it to setup/teardown console title.

File: test/test_networking.py
Comment (Grub4K) [reply test/test_networking.py:291]:
```suggestion
            self.assertIn(b'Accept-Encoding: identity', data)
            self.assertNotIn(b'youtubedl-no-compression', data.lower())
```
`youtubedl-no-compression` should not show up in the ouput anywhere, so this should be fine

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py:218]: this is not something that necessary needs to be done in the Request object - just somewhere in the underlying network stack before it is sent to the server (foreshadowing for network rework).

the test will fail if it encodes with lowercase since the test server is case sensitive

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py:254]: I didn't use a loop here to may it clear on the exact test cases we needed to cover and it wasn't too many

Comment (Grub4K) [reply test/test_networking.py:254]: I meant shouldnt we test that `HEAD`, `PUT`, `GET` also do not get changed?

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py:290]: that assumes that it is being replaced with accept encoding identity, which is what I should also be testing.

(this currently only tests that accept encoding identity is added and the header is not removed)

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py:254]: oh for 307/308? could do I suppose as an extra measure

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py:407]: technically could just pass `pathlib.Path(tf.name).as_uri()` which gives the same thing I said.

the request was a relic from network rework where it couldn't be a url :smile:

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py:361]: if we didn't use daemon = true wouldn't it block the setUp method when we start()?

But I suppose we could use tearDown to kill the thread - or better, move this out of setUp maybe so its global for all the test cases?

(this is legacy code from youtube-dl I mind you lol)

Comment (Grub4K) [reply test/test_networking.py:360]: > A thread can be flagged as a “[daemon](https://docs.python.org/3/library/threading.html#threading.Thread.daemon) thread”. The significance of this flag is that the entire Python program exits when only daemon threads are left. The initial value is inherited from the creating thread. The flag can be set through the daemon property or the daemon constructor argument.

`thread.start()` will not, it will only block if it is `join()`ed. So currently we are just spawning and destroying. If we set up state globally, we will have resources unmanaged by the test runner. I think its fine to keep it as is, remove the `daemon = True` and a `tearDown` method closing that thread.

File: test/test_networking.py
Comment (pukkandan) [inline test/test_networking.py:26]: Make the top of the file exactly:
```py
#!/usr/bin/env python3

# Allow direct execution
import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```

and `isort` the rest below this. I also prefer import `http.cookiejar` instead of `from`, but I'll leave that to your discretion.

**Review by pukkandan — APPROVED** [review]
I haven't compared the actual logic to the RFC - let me know if I should or you are confident enough to merge without.

While cleaning up legacy code certainly would be nice, I don't want to hold this PR up for that, so I'm not making any suggestions for it.

PS: It's difficult to see what's moved vs new code in test_networking. In future, try to do rename and code change in separate commits.

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py]: not gonna bother in this PR

File: test/test_networking.py
Comment (coletdjnz) [reply test/test_networking.py]: I did this but it started throwing unclosed socket warnings for some unknown reason in the test runner so I've reverted it for now. 

we can look into it at another time

Comment (Grub4K) [reply test/test_networking.py]: Sure

**Comment by dirkf** [comment]
In https://github.com/ytdl-org/youtube-dl/commit/c047270c02, yt-dl started to remove each `Content-Encoding` header after processing the encoding. The rationale was that the handler routine could be called twice with a geo-verification proxy set. If this still needs to be done, the tests `TestHTT.test_{encoding}` won't work as written. But shouldn't it be done anyway?

RFC9110 s8.4
>The "Content-Encoding" header field indicates what content codings have been applied to the representation, beyond those inherent in the media type, and thus what decoding mechanisms have to be applied in order to obtain data in the media type referenced by the Content-Type header field. ...

Obviously RFC9110 applies to what's sent over the network rather than the processing at either end, but it seems reasonable to apply the same semantics in those contexts, so that the header only lists the encodings that haven't yet been decoded. Work-arounds with a custom header or `Request` attribute seem unnecessarily painful.

**Comment by coletdjnz** [comment]
> In [ytdl-org/youtube-dl@c047270c02](https://github.com/ytdl-org/youtube-dl/commit/c047270c02), yt-dl started to remove each `Content-Encoding` header after processing the encoding. The rationale was that the handler routine could be called twice with a geo-verification proxy set. If this still needs to be done, the tests `TestHTT.test_{encoding}` won't work as written. But shouldn't it be done anyway?
> 
> RFC9110 s8.4
> 
> > The "Content-Encoding" header field indicates what content codings have been applied to the representation, beyond those inherent in the media type, and thus what decoding mechanisms have to be applied in order to obtain data in the media type referenced by the Content-Type header field. ...
> 
> Obviously RFC9110 applies to what's sent over the network rather than the processing at either end, but it seems reasonable to apply the same semantics in those contexts, so that the header only lists the encodings that haven't yet been decoded. Work-arounds with a custom header or `Request` attribute seem unnecessarily painful.

This was changed in https://github.com/yt-dlp/yt-dlp/commit/65e5c021e7c5f23ecbc6a982b72a02ac6cd6900d.

Funny enough, even urllib3/requests don't remove the Content-Encoding header, hence why I gave the go ahead.

I'll have to test the geo verification/per request proxy. I completely changed it in the network rework as the implementation is flawed/buggy.

**Comment by dirkf** [comment]
And I actually commented on that PR too. But I hadn't yet tracked down the commit where the removal was added.

Looking at the MDN page for content-encoding (tl;dr for the RFCs), the decoding logic is indeed essentially `for encoding in reversed(headers['Content-Encoding'].split()): data = decode(data, encoding)`. However nested encoding would appear to have little benefit with the standard set of encodings, and presumably would only happen accidentally in some proxy configuration.

**Comment by dirkf** [comment]
Reading the HTTP specs again, it seems that a header (like `Content-encoding`) that can have multiple (comma-separated) values can also appear as multiple lines each containing one or more of the same set of values, so these should be equivalent:
```
Multi-valued-hdr: val1, val2, val3
```
```
Multi-valued-hdr: val1, val2
...
Multi-valued-hdr: val3
```
So my code for this is (where the decompression methods return `io.BytesIO()`) and `txt_or_none()` is `lambda x: strip_or_none(x) or None`:
```py
        encodings = list(filter(txt_or_none, ', '.join(resp.headers.get_all('Content-encoding', [])).split(',')))
        for enc in encodings[::-1]:
            old_resp = resp
            if enc in ('gzip', 'x-gzip'):
                uncompressed = self.gzip_d(resp.read())
                encodings.pop()
            elif enc == 'deflate':
                uncompressed = self.reflate(resp.read())
                encodings.pop()
            elif enc == 'br' and brotli:
                uncompressed = self.brotli(resp.read())
                encodings.pop()
            else:
                break
            resp = compat_urllib_request.addinfourl(
                uncompressed, old_resp.headers, old_resp.url, old_resp.code)
            resp.msg = old_resp.msg

        if encodings:
            # remaining encodings
            resp.headers['Content-encoding'] = ' '.join(encodings)
        elif 'Content-encoding' in resp.headers:
            del resp.headers['Content-encoding']
```

**Pull request merged**
